### PR TITLE
add `QueryMultiDict`

### PR DESCRIPTION
**Pull request by Goldziher**
# PR Checklist

- [x] Have you followed the guidelines in `CONTRIBUTING.md`?
- [x] Have you got 100% test coverage on new code?
- [x] Have you updated the prose documentation?
- [x] Have you updated the reference documentation?

This PR adds a multi dict for query params. 

It might be that this implementation will degrade performance a bit, so we need to test and see how it stacks vis-a-vis the previous one. I will run some benchmarks tomorrow to see.

### CODE REVIEW

**Review by peterschutt — APPROVED** [review]

**Comment by provinzkraut** [comment]
@Goldziher Let me know what the outcome of your tests is. If it looks good, I'll incorporate it into #667

File: starlite/datastructures/multi_dicts.py [hunk 1]
```diff
@@ -41,3 +42,41 @@ async def close(self) -> None:
...
+
+        Returns:
+            A QueryMultiDict instance
+        """
+        _bools = {"true": True, "false": False, "True": True, "False": False}
+        return cls(
+            (k, v) if v not in _bools else (k, _bools[v]) for k, v in parse_qsl(query_string, keep_blank_values=True)
+        )
+
+    def dict(self) -> Dict[str, List[Any]]:
```
Comment (provinzkraut) [inline starlite/datastructures/multi_dicts.py:68]: Why not

```suggestion
    def dict(self) -> Dict[str, List[Any]]:
        return {self.getall(key) for key in set(self.keys)}`
```

File: starlite/datastructures/multi_dicts.py [hunk 2]
```diff
@@ -41,3 +42,41 @@ async def close(self) -> None:
         for _, value in self.multi_items():
             if isinstance(value, UploadFile):
                 await value.close()
+
+
+class QueryMultiDict(MultiDict[Any]):
```
Comment (provinzkraut) [inline starlite/datastructures/multi_dicts.py:47]: The way we parse it, the items can only ever be a `str` or `bool`, so why not annotate it as such?

File: starlite/datastructures/multi_dicts.py (hunk 1 above)
Comment (Goldziher) [reply starlite/datastructures/multi_dicts.py:68]: i would assume this is slower, ill test

**Pull request merged**